ouroboros-consensus-protocol/src/unstable-protocol-testlib/Test/Ouroboros/Consensus/Protocol/Praos/Header.hs

```diff
@@ -0,0 +1,463 @@
+                let rho' = mkInputVRF newSlotNo nonce
+                    period' = unSlotNo newSlotNo `div` praosSlotsPerKESPeriod
+                    hbVrfRes = VRF.evalCertified () rho' vrfSignKey
+                    newBody = body{hbSlotNo = newSlotNo, hbVrfRes}
+                    sig' = KES.signKES () (fromIntegral period' - kesPeriod) newBody kesSignKey
+                pure (context, Header newBody (KES.SignedKES sig'))
+            MutateCounterOver1 -> do
+                let poolId = coerce $ hashKey $ VKey $ deriveVerKeyDSIGN coldSignKey
+                    Header body _ = header
+                    OCert{ocertN} = hbOCert body
+                newCounter <- choose (0, ocertN - 2)
+                let context' = context{ocertCounters = Map.insert poolId newCounter (ocertCounters context)}
+                pure (context', header)
+            MutateCounterUnder -> do
+                let poolId = coerce $ hashKey $ VKey $ deriveVerKeyDSIGN coldSignKey
+                    oldCounter = fromMaybe 0 $ Map.lookup poolId (ocertCounters context)
+                newCounter <- arbitrary `suchThat` (> oldCounter)
+                let context' = context{ocertCounters = Map.insert poolId newCounter (ocertCounters context)}
+                pure (context', header)
+    GeneratorContext{praosSlotsPerKESPeriod, praosMaxKESEvo, kesSignKey, vrfSignKey, coldSignKey, nonce} = context
+
+data Mutation
+    = -- | No mutation
+      NoMutation
+    | -- | Mutate the KES key, ie. sign the header with a different KES key.
+      MutateKESKey
+    | -- | Mutate the cold key, ie. sign the operational certificate with a different cold key.
+      MutateColdKey
+    | -- | Mutate the KES period in the operational certificate to be
+      -- after the start of the KES period.
+      MutateKESPeriod
+    | -- | Mutate KES period to be before the current KES period
+      MutateKESPeriodBefore
+    | -- | Mutate certificate counter to be greater than expected
+      MutateCounterOver1
+    | -- | Mutate certificate counter to be lower than expected
+      MutateCounterUnder
+    deriving (Eq, Show)
+
+instance Json.ToJSON Mutation where
+    toJSON = \case
+        NoMutation -> "NoMutation"
+        MutateKESKey -> "MutateKESKey"
+        MutateColdKey -> "MutateColdKey"
+        MutateKESPeriod -> "MutateKESPeriod"
+        MutateKESPeriodBefore -> "MutateKESPeriodBefore"
+        MutateCounterOver1 -> "MutateCounterOver1"
+        MutateCounterUnder -> "MutateCounterUnder"
+
+instance Json.FromJSON Mutation where
+    parseJSON = \case
+        "NoMutation" -> pure NoMutation
+        "MutateKESKey" -> pure MutateKESKey
+        "MutateColdKey" -> pure MutateColdKey
+        "MutateKESPeriod" -> pure MutateKESPeriod
+        "MutateKESPeriodBefore" -> pure MutateKESPeriodBefore
+        "MutateCounterOver1" -> pure MutateCounterOver1
+        "MutateCounterUnder" -> pure MutateCounterUnder
+        _ -> fail "Invalid mutation"
```

Comment:
These two seem very standard, doesn't deriving the instances yield the same result?